app/models/highlight.rb

```diff
@@ -16,54 +21,113 @@ class Highlight < ApplicationRecord
                         :highlighted_content,
                         :location_strategies
 
-  validate :source_ids_uuid_for_openstax_page
+  validate :must_have_source_type_specific_ids
+  validate :must_have_consistent_neighbors, on: :create
 
   enum source_type: [:openstax_page]
 
-  before_validation :normalize_color, :normalize_source_ids
-
   scope :by_source_type, ->(source_type) { where(source_type: source_type) }
   scope :by_color, ->(color) { where(color: color) }
   scope :by_user, ->(user_id) { where(user_uuid: user_id) }
-  scope :by_order, ->(order_direction) {
-    order_direction = order_direction.to_sym
-    order(order_in_source: order_direction, source_order: order_direction)
-  }
-  scope :by_source_parent_ids, ->(source_parent_ids) do
-    where('source_parent_ids && ?', postgres_style_array(source_parent_ids))
-  end
+  scope :by_scope_id, ->(scope_id) { where(scope_id: scope_id) }
+  scope :by_source_ids, ->(source_ids) { where(source_id: source_ids) }
+
+  around_create :insert_new_highlight_between_neighbors
```

Comment:
we've already discussed this and i know your opinion, just restating my opinion here for the record.   it's really good to avoid AR callbacks http://codeatmorning.com/3-reasons-to-avoid-rails-callbacks/.  This app will grow, tests will continue to be added, more model associations will be added....these sorts of hooks are 99% guaranteed to cause headaches in the future.    

A good alternative to them, imo, is to create a higher level abstraction:
```
     HighlightNeighborConnector
        def create 
            set_neighbors(highlight)
            highlight.create
       end

      def destroy(highlight)
          reconnect_neighbors
          highlight.destroy
     end
```

Reply:
I generally agree with you which is why I wrote https://github.com/lml/lev#routines (a routine would be like the above except with transactions built in).  The exception in my mind is when a callback only works inside one model.  Mostly I'm ok with this with a model affecting its own state.  I admit that this `Highlight` case is a little different because it is impacting its own state and the state of other highlight models.  So it is borderline for me, in this case I'm ok with it b/c I see Highlight as a linked list model so having the glue in there for linked listing is ok